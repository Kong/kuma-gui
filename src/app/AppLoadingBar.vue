<template>
  <div class="full-screen">
    <div class="loading-container">
      <img src="@/assets/images/kuma-loader-v1.gif">

      <div class="progress">
        <div
          :style="{ width: `${progress}%` }"
          class="progress-bar"
          role="progressbar"
          data-testid="app-progress-bar"
        />
      </div>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref, onMounted, onUnmounted } from 'vue'

let timer: number
const progress = ref<number>(10)

onMounted(function () {
  timer = window.setInterval(() => {
    if (progress.value >= 100) {
      window.clearInterval(timer)
      progress.value = 100
    }

    progress.value = Math.min(progress.value + Math.ceil(Math.random() * 30), 100)
  }, 150)
})

onUnmounted(function () {
  window.clearInterval(timer)
})
</script>

<style lang="scss" scoped>
.full-screen {
  background: var(--white);
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  z-index: 50000;
  display: flex;
  align-items: center;
  justify-content: center;
}

.loading-container {
  display: flex;
  max-width: 350px;
  align-items: center;
  flex-direction: column;

  .progress {
    margin-top: 1rem;
    width: 130%;
    background-color: var(--grey-300);

    .progress-bar {
      height: 5px;
      background-color: var(--blue-400);
    }
  }
}
</style>
