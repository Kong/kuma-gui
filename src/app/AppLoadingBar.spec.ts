import { afterAll, beforeAll, describe, expect, jest, test } from '@jest/globals'
import { flushPromises, shallowMount } from '@vue/test-utils'

import AppLoadingBar from './AppLoadingBar.vue'

function renderComponent() {
  return shallowMount(AppLoadingBar)
}

describe('AppLoadingBar', () => {
  beforeAll(() => {
    jest.useFakeTimers()
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  test('renders snapshot at the beginning', () => {
    const wrapper = renderComponent()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('checks if width is max of 100% size', async () => {
    const wrapper = renderComponent()

    jest.advanceTimersByTime(1500)
    await flushPromises()

    expect(wrapper.find('[data-testid="app-progress-bar"]').attributes('style')).toBe('width: 100%;')
  })
})
