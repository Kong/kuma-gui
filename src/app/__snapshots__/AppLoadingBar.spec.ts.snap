// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppLoadingBar renders snapshot at the beginning 1`] = `
<div>
  <div
    class="full-screen"
  >
    <div
      class="loading-container"
    >
      <img />
      <div
        class="progress"
      >
        <div
          class="progress-bar"
          role="progressbar"
          style="width: 10%;"
        />
      </div>
    </div>
  </div>
</div>
`;
